Implemented task "Goal setting" BJS2-53998

MykhailoShpakov07 opened the pull request.

MykhailoShpakov07 committed:
Implemented task "Goal setting" BJS2-53998

MykhailoShpakov07 commented:
https://faang-school.atlassian.net/browse/BJS2-53998

porchy21 commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,134 @@
+                .toList();
+    }
+
+    private void setAllMissingFields(Goal goalTo, Goal goalFrom) {
+        goalTo.setId(goalTo.getId() == null ? goalFrom.getId() : goalTo.getId());
+        goalTo.setParent(goalTo.getParent() == null ? goalFrom.getParent() : goalTo.getParent());
```
Выглядит жутко и неподдерживаемым, однако я осознаю необходимость этого метода. Ты сравнивал этот подход и способ  с использованием Java Reflection Api для пробежки по всем полям класса Goal через цикл? Кстати, а если будет логика обнулить поле, к примеру убрать дедлайн, сделать это не получится.

20cognosce replied:
приватные методы внизу, публичные вверху

выглядит не очень, но сразу идей как улучшить не придумал
надо пофантазировать над рефакторингом этого метода

20cognosce replied:
рефлексию не надо использовать точно

porchy21 commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,134 @@
+        return goalUpdated;
+    }
+
+    @Transactional
+    public void deleteGoal(Long goalId) {
+        goalRepository.deleteById(goalId);
+    }
```
Для чего методу, состоящему из одной строки `@Transactional`?

MykhailoShpakov07 replied:
Ты прав, поскольку метод состоит из всего лишь одной инструкции, транзакция не обязательна

porchy21 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,88 @@
+            @RequestBody GoalFilterDto goalFilterDto) {
+
+
+        List<GoalDto> goalDtoList = goalService.findSubGoalsByParentId(parentGoalId, goalFilterDto).stream()
+                .map(goalMapper::toDto).toList();
```
Писал также, Дима (ревьюер) сказал вынести результат в переменную и только потом маппить. К тому же можно же написать метод toDtoList в маппере, вместо того чтобы лепить стримы

DmitryQueen replied:
Можно просто в маппере написать метод toDtoList который принимает лист сущностей и мапстракт сам поймет что нужно лист смаппить и вернет лист дтошек

Я так сделал:
`    List<GoalInvitationDto> toDtoList(List<GoalInvitation> goalInvitations);
`

MykhailoShpakov07 replied:
Спасибо за подсказку. Исправлю !

porchy21 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,88 @@
+            @PathVariable @Positive(message = "Please, provide positive user ID") Long userId,
+            @RequestBody @Valid GoalDto goalDto) {
+
+        log.info("Creating goal for user with id {} and goalDto {}", userId, goalDto);
```
Как мне кажется, избыточно и не нужно.

MykhailoShpakov07 replied:
Ты имеешь ввиду написание логов или валидация входных данных с помощью @Positive и @Valid ? Я считаю что нужно поскольку это семантическая валидация не связаная с бизнес логикой и ее следует выполнять в контроллере или классе валидаторе

porchy21 replied:
> Ты имеешь ввиду написание логов или валидация входных данных с помощью @positive и @Valid ? Я считаю что нужно поскольку это семантическая валидация не связаная с бизнес логикой и ее следует выполнять в контроллере или классе валидаторе

Немного нафидил при ревью, я имел ввиду логирование подобным образом, про валидацию в точку

porchy21 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,88 @@
+    private final GoalService goalService;
+    private final GoalMapper goalMapper;
+
+    @PostMapping("/create/{userId}")
```
URL некорректны: зачем лишний раз писать create/update/delete если для их вызова из сети уже используются соответствующие HTTP-операции POST, PATCH, DELETE.

MykhailoShpakov07 replied:
Мне кажеться это больше дело вкуса. Если использовать разные url для операций то endpoint'ы тогда проще отличать между собой и меньше шанса напутать что-то, поэтому не считаю ошибкой добавление другого url под соответствуещий http метод

20cognosce replied:
в адресе эндпоинта не должно быть действия, так как за это отвечает метод http
получается масло масленное

porchy21 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,88 @@
+                body(goalMapper.toDto(createdGoal));
+    }
+
+    @PutMapping("/update/{goalId}")
```
Ты же обновляешь только измененные поля, Patch

MykhailoShpakov07 replied:
Да, можно и Patch

porchy21 commented on src/test/java/school/faang/user_service/controller/goal/GoalControllerTest.java:
```diff
@@ -0,0 +1,124 @@
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.when;
+
+class GoalControllerTest {
```
Не знаю насколько нужен этот тест. На встречах обсуждали, что тесты пишутся для всего, что имеют бизнес-логику: сервисы, валидаторы и т.д

20cognosce replied:
На контроллер пишутся интеграционные тесты с отправкой локального запроса по сети, юнит для него не нужен, так как в нем не должно быть логики

porchy21 commented on src/test/java/school/faang/user_service/filter/goal/StatusFilterTest.java:
```diff
@@ -0,0 +1,51 @@
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class StatusFilterTest {
```
Тесты для фильтров тоже выглядят странно, мне кажется достаточно просто в тесте сервиса протестировать методы и с их использованием.

MykhailoShpakov07 replied:
Не согласен. В тестах сервиса необходимо тестировать только сервис и все его зависимости должны быть mock'нутые ( включая фильтры ). Поэтому я создал отдельные юнит-тесты под каждый фильтр что-бы проверить их корректное функционирование применяя их к одному filterDto ( этого достаточного )

porchy21 reviewed (commented):
В целом это отличная масштабная работа, есть только небольшие придирки по написанию кода. Тестам уделил меньше внимания, надеюсь там все в порядке

DmitryQueen commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,88 @@
+            @RequestBody GoalFilterDto goalFilterDto) {
+
+        List<GoalDto> goalDtoList = goalService.findSubGoalsByUserId(userId, goalFilterDto).stream()
+                .map(goalMapper::toDto).toList();
```
тут такая же ситуация можно в маппере метод сделать для смапливания листов

MykhailoShpakov07 replied:
Исправлю. Спасибо !

DmitryQueen commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,134 @@
+
+    @Transactional
+    public Goal createGoal(Long userId, String title, String description, Long parentId, List<Long> skillIds) {
+        //validation
```
комменты убираем

DmitryQueen commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,134 @@
+    @Transactional
+    private void assignSkillsToGoal(Long goalId, List<Long> skillsId) {
+        if (skillsId == null) {
+            return;
```
было бы норм сделать какое-то логирование, не просто же так метод прервался

20cognosce replied:
Ошибку надо кидать, а не пустой return

MykhailoShpakov07 committed:
Corrected task as suggested.
Introduced additional checks in createGoal and findSubGoalsByUserId and updateGoal methods of GoalService

20cognosce commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,82 @@
+            @RequestBody @Valid GoalDto goalDto) {
+
+        Goal createdGoal = goalService.createGoal(userId, goalDto.getTitle(), goalDto.getDescription(), goalDto.getParentId(), goalDto.getSkillIds());
+        return ResponseEntity.status(HttpStatus.CREATED).
+                body(goalMapper.toDto(createdGoal));
```
не надо слепливать, переменные объявляй

20cognosce commented on src/main/java/school/faang/user_service/mapper/GoalMapper.java:
```diff
@@ -0,0 +1,32 @@
+    @Mapping(source = "skillsToAchieve", target = "skillIds", qualifiedByName = "mapSkills")
+    GoalDto toDto(Goal entity);
+
+    @Mapping(source = "parentId", target = "parent.id")
```
не мапь айдишники на сущности, передавай их в сервис

20cognosce commented on src/main/java/school/faang/user_service/repository/goal/GoalRepository.java:
```diff
@@ -47,4 +43,19 @@ WITH RECURSIVE subtasks AS (
             """)
     List<User> findUsersByGoalId(long goalId);
+
+    @Transactional
```
не в репозитории, а в сервисе использовать аннотацию транзакции

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,158 @@
+        Stream<Goal> subGoals = goalRepository.findGoalsByUserId(userId);
+        List<GoalFilter> applicableFilters = goalFilters.stream().filter(goalFilter -> goalFilter.isApplicable(filterDto)).toList();
+
+        return subGoals.filter(goal ->
+                        applicableFilters.stream().allMatch(filter -> filter.apply(filterDto, goal))
+                )
+                .toList();
```
что с форматированием

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,158 @@
+            throw new NoSuchElementException(String.format("User with id %s doesn't exist", userId));
+        }
+
+        //perform retrieval and filtration
```
не писать бессмысленные комментарии

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,158 @@
+
+        //perform retrieval and filtration
+        Stream<Goal> subGoals = goalRepository.findGoalsByUserId(userId);
+        List<GoalFilter> applicableFilters = goalFilters.stream().filter(goalFilter -> goalFilter.isApplicable(filterDto)).toList();
```
почему стримы в одну строчку

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,158 @@
+    @Transactional(readOnly = true)
+    public List<Goal> findSubGoalsByParentId(Long parentGoalId, GoalFilterDto filterDto) {
+        Stream<Goal> subGoals = goalRepository.findByParent(parentGoalId);
+        List<GoalFilter> applicableFilters = goalFilters.stream().filter(goalFilter -> goalFilter.isApplicable(filterDto)).toList();
+
+        return subGoals.filter(goal ->
+                        applicableFilters.stream().allMatch(filter -> filter.apply(filterDto, goal))
+                )
+                .toList();
```
.

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,158 @@
+            log.info("Skills with ids {} have been set for the goal with id {}", skillIds, goalId);
+        }
+
+        if (Objects.equals(goalUpdated.getStatus(), GoalStatus.COMPLETED)) {
```
енамы можно через ==

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,158 @@
+    @Transactional
+    public Goal updateGoal(Long goalId, Goal goalUpdated, List<Long> skillIds) {
+        //validation
+        Goal goalOld = goalRepository.findById(goalId).orElseThrow(() -> new NoSuchElementException(String.format("No goal found with such id %s", goalId)));
```
почему слеплено в одну строку

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,158 @@
+@Slf4j
+@RequiredArgsConstructor
+@Service
+public class GoalService {
```
сделать методы чище

MykhailoShpakov07 committed:
Corrected task as suggested by the mentor
